activesupport/lib/active_support/callbacks.rb

```diff
@@ -400,21 +409,15 @@ def initialize(target, method, arguments, block)
         # The actual invocation is left up to the caller to minimize
         # call stack pollution.
         def expand(target, value, block)
-          result = @arguments.map { |arg|
-            case arg
-            when :value; value
-            when :target; target
-            when :block; block || raise(ArgumentError)
-            end
-          }
-
-          result.unshift @method_name
-          result.unshift @override_block || block
-          result.unshift @override_target || target
+          base_result = [@override_target || target, @override_block || block, @method_name]
 
-          # target, block, method, *arguments = result
-          # target.send(method, *arguments, &block)
-          result
+          @arguments.each_with_object(base_result) do |arg, result|
+            result << case arg
```

Comment:
What purpose is this change?

Reply:
Again, it is merely a performance suggestion and we can simply drop it if it's not desired. Based on the benchmarks I explored, it is faster to declare the array and then concatenate the other arguments than unshifting it three times.

Additionally, it is arguably more readable.

Reply:
I'm not liking the specific style here, any way we can do something like this?

```ruby
expanded = [ @override_target || target, @override_block || block, @method_name ]

@arguments.each do |argument|
  case argument
  when :value  then expanded << value
  when :target then expanded << target
  when :block  then expanded << block
  else raise ArgumentError
  end
end
```

Reply:
Another potential shot:

```ruby
values    = { target: target, value: value, block: block }
arguments = @arguments.map { |arg| values[arg] || raise ArgumentError }

[ @override_target || target, @override_block || block, @method_name, *arguments ]
```

Not necessarily what we should go with, but think there's a version lurking here that's clearer but still more performant. For instance, we own all the `expand` callers, so could be worth trying a keyword argument version too.

Reply:
Great suggestions. I'll explore those ideas further and try to come up with something cleaner.

Reply:
I have tried a few implementations and - mostly because the `ArgumentError` should only be raised if the argument type is block and the block value is empty - the cleanest one that still delivers a performance benefit is the current one. Using kwargs was very significantly slower.